src/Control/Distributed/Process/Internal/CQueue.hs

```diff
@@ -101,33 +106,58 @@ dequeue :: forall m a.
         -> BlockSpec         -- ^ Blocking behaviour
         -> [MatchOn m a]     -- ^ List of matches
         -> IO (Maybe a)      -- ^ 'Nothing' only on timeout
-dequeue (CQueue arrived incoming) blockSpec matchons = mask_ $
+dequeue (CQueue arrived incoming size) blockSpec matchons =
   case blockSpec of
-    Timeout n -> timeout n $ fmap fromJust run
+    Timeout n -> registerDelay n >>= run
     _other    ->
        case chunks of
          [Right ports] -> -- channels only, this is easy:
            case blockSpec of
-             NonBlocking -> atomically $ waitChans ports (return Nothing)
-             _           -> atomically $ waitChans ports retry
+             NonBlocking -> atomically $ decrementJust $ waitChans ports (return Nothing)
+             _           -> atomically $ decrementJust $ waitChans ports retry
                               -- no onException needed
-         _other -> run
+         _other -> newTVarIO False >>= run
   where
     chunks = chunkMatches matchons
 
-    run = do
-           arr <- takeMVar arrived
-           let grabNew xs = do
-                 r <- atomically $ tryReadTChan incoming
-                 case r of
-                   Nothing -> return xs
-                   Just x  -> grabNew (Snoc xs x)
-           arr' <- grabNew arr
-           goCheck chunks arr'
+    decrementJust f = do
+      mx <- f
+      when (isJust mx) (modifyTVar' size pred)
+      return mx
 
     waitChans ports on_block =
         foldr orElse on_block (map (fmap Just) ports)
 
+    run :: TVar Bool -> IO (Maybe a)
```

Comment:
It should decrement queue size if value is read, and do nothing orderwise (I'm not sure if this PR worth reviewing, as anyway we have no benchmarks to understand if it's better than IO based or not, I think I should just close it until discussion will happen again)